src/compiler/checker.ts

```diff
@@ -7139,6 +7139,42 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             }
 
             const expandedParams = getExpandedParameters(signature, /*skipUnionExpanding*/ true)[0];
+
+            // For regular function/method declarations, the enclosing declaration will already be signature.declaration,
+            // so this is a no-op, but for arrow functions and function expressions, the enclosing declaration will be
+            // the declaration that the arrow function / function expression is assigned to.
+            //
+            // If the parameters or return type include "typeof globalThis.paramName", using the wrong scope will lead
+            // us to believe that we can emit "typeof paramName" instead, even though that would refer to the parameter,
+            // not the global. Make sure we are in the right scope by changing the enclosingDeclaration to the function.
+            //
+            // We can't use the declaration directly; it may be in another file and so we may lose access to symbols
+            // accessible to the current enclosing declaration, or gain access to symbols not accessible to the current
+            // enclosing declaration. To keep this chain accurate, clone the function as a non-synthesized node (so that
+            // it can be used as an enclosing declaration) and make its parent the current enclosing declaration.
+            //
+            // If the declaration is in a JS file, then we don't need to do this at all, as there are no annotations besides
+            // JSDoc, which are always outside the function declaration, so are not in the parameter scope.
+            const saveEnclosingDeclaration = context.enclosingDeclaration;
+            if (saveEnclosingDeclaration && signature.declaration && signature.declaration !== saveEnclosingDeclaration && !isInJSFile(signature.declaration)) {
+                const fakeScope = parseNodeFactory.createFunctionExpression(

```

Comment:
While this method does work, it has the downside of introducing a synthetic node into the declaration chain, and so when this happens, the emit suffers. I'm not sure of a better way; `isSymbolAccessible` currently seems to only use the enclosing declaration and its parents, so I don't know how to get this info into its scope.